File: local_proxy.py
import json
import os
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import requests


BASE_API = "https://api.cubemaster.net"
HOST = "127.0.0.1"
PORT = 8000
WEB_FILE = Path(__file__).with_name("frontend_mock.html")


def pick_token(query: dict[str, list[str]], body: dict | None, headers) -> str:
    token = ""
    if "token" in query and query["token"]:
        token = query["token"][0]
    if not token and isinstance(body, dict):
        token = str(body.get("token", ""))
    if not token:
        token = headers.get("X-TokenID", "")
    if not token:
        token = os.getenv("CUBEMASTER_TOKEN", "")
    return token


def forward(method: str, path: str, token: str, params=None, payload=None) -> tuple[int, dict | str]:
    if not token:
        return HTTPStatus.BAD_REQUEST, {"error": "Missing token. Provide token query param, X-TokenID header, or CUBEMASTER_TOKEN env var."}

    headers = {"TokenID": token}
    response = requests.request(
        method=method,
        url=f"{BASE_API}{path}",
        headers=headers,
        params=params,
        json=payload,
        timeout=40,
    )

    content_type = response.headers.get("Content-Type", "")
    if "application/json" in content_type:
        try:
            return response.status_code, response.json()
        except ValueError:
            return response.status_code, {"raw": response.text}
    return response.status_code, response.text


class ProxyHandler(BaseHTTPRequestHandler):
    def _send_json(self, status: int, payload):
        body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, X-TokenID")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.end_headers()
        self.wfile.write(body)

    def _send_html(self, html: str):
        data = html.encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, X-TokenID")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        query = parse_qs(parsed.query)

        if parsed.path == "/" or parsed.path == "/index.html":
            if WEB_FILE.exists():
                self._send_html(WEB_FILE.read_text(encoding="utf-8"))
            else:
                self._send_html("<h1>frontend_mock.html not found</h1>")
            return

        if parsed.path == "/api/health":
            self._send_json(200, {"ok": True, "api": BASE_API})
            return

        if parsed.path == "/api/check-token":
            token = pick_token(query, None, self.headers)
            limit = int(query.get("limit", ["1"])[0])
            status, body = forward("GET", "/Loads", token, params={"limit": limit})
            self._send_json(status, body)
            return

        if parsed.path == "/api/loads":
            token = pick_token(query, None, self.headers)
            limit = int(query.get("limit", ["1"])[0])
            status, body = forward("GET", "/Loads", token, params={"limit": limit})
            self._send_json(status, body)
            return

        if parsed.path.startswith("/api/loads/"):
            token = pick_token(query, None, self.headers)
            load_id = parsed.path.split("/api/loads/", 1)[1]
            status, body = forward("GET", f"/Loads/{load_id}", token)
            self._send_json(status, body)
            return

        self._send_json(404, {"error": "Not Found"})

    def do_POST(self):
        parsed = urlparse(self.path)
        content_len = int(self.headers.get("Content-Length", "0"))
        raw_body = self.rfile.read(content_len) if content_len else b"{}"

        try:
            payload = json.loads(raw_body.decode("utf-8"))
        except json.JSONDecodeError:
            self._send_json(400, {"error": "Invalid JSON body"})
            return

        query = parse_qs(parsed.query)

        if parsed.path == "/api/loads":
            token = pick_token(query, payload, self.headers)
            status, body = forward("POST", "/Loads", token, payload=payload)
            self._send_json(status, body)
            return

        self._send_json(404, {"error": "Not Found"})


if __name__ == "__main__":
    server = ThreadingHTTPServer((HOST, PORT), ProxyHandler)
    print(f"Local proxy running at http://{HOST}:{PORT}")
    print("Endpoints: /api/health, /api/check-token, /api/loads, /api/loads/{id}")
    server.serve_forever()

File: local_proxy.js
import { createServer } from 'http';
import { readFileSync, existsSync } from 'fs';
import { resolve, dirname, extname } from 'path';
import { fileURLToPath } from 'url';
import mysql from 'mysql2/promise';

const BASE_API = 'https://api.cubemaster.net';
const HOST = '0.0.0.0';
const PORT = 3001;
const SERVER_TOKEN = process.env.CUBEMASTER_TOKEN;

function requireEnv(name) {
  const value = process.env[name];
  if (!value) {
    throw new Error(`[CONFIG] Missing required environment variable: ${name}`);
  }
  return value;
}

const MYSQL_HOST = process.env.MYSQL_HOST || '127.0.0.1';
const MYSQL_PORT = parseInt(process.env.MYSQL_PORT || '3306', 10);
const MYSQL_USER = process.env.MYSQL_USER || 'root';
const MYSQL_PASSWORD = requireEnv('MYSQL_PASSWORD');
const MYSQL_DATABASE = process.env.MYSQL_DATABASE || 'cubemaster';
const __dirname = dirname(fileURLToPath(import.meta.url));
const WEB_FILE = resolve(__dirname, 'frontend_mock.html');

const dbPool = mysql.createPool({
  host: MYSQL_HOST,
  port: MYSQL_PORT,
  user: MYSQL_USER,
  password: MYSQL_PASSWORD,
  database: MYSQL_DATABASE,
  waitForConnections: true,
  connectionLimit: 10,
  queueLimit: 0,
});

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type, X-TokenID',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
};

function pickToken() {
  return SERVER_TOKEN;
}

async function logPalletLoad(inputPayload, outputBody, responseStatus) {
  const inputJson = JSON.stringify(inputPayload ?? {});
  const outputJson = JSON.stringify(outputBody ?? null);
  await dbPool.execute(
    `INSERT INTO pallet_load_logs (input_json, output_json, response_status)
     VALUES (?, ?, ?)`,
    [inputJson, outputJson, responseStatus ?? null]
  );
}

async function logPallet2CtnLoad(flowId, step, inputPayload, outputBody, responseStatus) {
  if (!flowId) return;
  const inputJson = JSON.stringify(inputPayload ?? {});
  const outputJson = JSON.stringify(outputBody ?? null);
  if (String(step) === '1') {
    await dbPool.execute(
      `INSERT INTO pallet2ctn_load_logs (flow_id, step1_input_json, step1_output_json, step1_status)
       VALUES (?, ?, ?, ?)
       ON DUPLICATE KEY UPDATE
         step1_input_json = VALUES(step1_input_json),
         step1_output_json = VALUES(step1_output_json),
         step1_status = VALUES(step1_status),
         updated_at = CURRENT_TIMESTAMP`,
      [flowId, inputJson, outputJson, responseStatus ?? null]
    );
  } else if (String(step) === '2') {
    await dbPool.execute(
      `INSERT INTO pallet2ctn_load_logs (flow_id, step2_input_json, step2_output_json, step2_status)
       VALUES (?, ?, ?, ?)
       ON DUPLICATE KEY UPDATE
         step2_input_json = VALUES(step2_input_json),
         step2_output_json = VALUES(step2_output_json),
         step2_status = VALUES(step2_status),
         updated_at = CURRENT_TIMESTAMP`,
      [flowId, inputJson, outputJson, responseStatus ?? null]
    );
  }
}

async function logLoadByFlow(query, inputPayload, outputBody, responseStatus) {
  const flowType = query.get('flowType') || '';
  if (flowType === 'pallet') {
    await logPalletLoad(inputPayload, outputBody, responseStatus);
    return;
  }
  if (flowType === 'pallet2ctn') {
    await logPallet2CtnLoad(
      query.get('flowId') || '',
      query.get('step') || '',
      inputPayload,
      outputBody,
      responseStatus
    );
  }
}

async function forward(method, path, token, { params, payload } = {}) {
  if (!token) {
    return {
      status: 400,
      body: { error: 'Missing token. Provide token query param, X-TokenID header, or CUBEMASTER_TOKEN env var.' },
    };
  }
  let url = `${BASE_API}${path}`;
  if (params) {
    url += `?${new URLSearchParams(params).toString()}`;
  }
  const init = {
    method,
    headers: { 'TokenID': token },
    signal: AbortSignal.timeout(180000),
  };
  if (payload !== undefined) {
    init.headers['Content-Type'] = 'application/json';
    init.body = JSON.stringify(payload);
    console.log('[→ CubeMaster Request]', url);
    console.log(JSON.stringify(payload, null, 2));
  }
  const resp = await fetch(url, init);
  const contentType = resp.headers.get('content-type') || '';
  if (contentType.includes('application/json')) {
    try {
      return { status: resp.status, body: await resp.json() };
    } catch {
      return { status: resp.status, body: { raw: await resp.text() } };
    }
  }
  return { status: resp.status, body: await resp.text() };
}

function sendJson(res, status, payload) {
  const body = Buffer.from(JSON.stringify(payload), 'utf-8');
  res.writeHead(status, {
    'Content-Type': 'application/json; charset=utf-8',
    'Content-Length': body.length,
    ...CORS_HEADERS,
  });
  res.end(body);
}

function sendHtml(res, html) {
  const body = Buffer.from(html, 'utf-8');
  res.writeHead(200, {
    'Content-Type': 'text/html; charset=utf-8',
    'Content-Length': body.length,
    'Cache-Control': 'no-store',
  });
  res.end(body);
}

const MIME_TYPES = {
  '.html': 'text/html',
  '.css':  'text/css',
  '.js':   'application/javascript',
  '.json': 'application/json',
  '.png':  'image/png',
  '.jpg':  'image/jpeg',
  '.svg':  'image/svg+xml',
};

function serveStatic(res, filePath) {
  const safePath = resolve(__dirname, filePath);
  if (!safePath.startsWith(__dirname)) {
    sendJson(res, 403, { error: 'Forbidden' });
    return;
  }
  if (!existsSync(safePath)) {
    sendJson(res, 404, { error: 'Not Found' });
    return;
  }
  const ext = extname(safePath).toLowerCase();
  const mime = MIME_TYPES[ext] || 'application/octet-stream';
  const body = readFileSync(safePath);
  res.writeHead(200, {
    'Content-Type': mime + (mime.startsWith('text') ? '; charset=utf-8' : ''),
    'Content-Length': body.length,
    'Cache-Control': 'no-store',
  });
  res.end(body);
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    let data = '';
    req.on('data', chunk => { data += chunk; });
    req.on('end', () => {
      try { resolve(data ? JSON.parse(data) : {}); }
      catch { reject(new Error('Invalid JSON')); }
    });
    req.on('error', reject);
  });
}

const server = createServer(async (req, res) => {
  const parsed = new URL(req.url, `http://${req.headers.host || HOST}`);
  const pathname = parsed.pathname;
  const query = parsed.searchParams;

  if (req.method === 'OPTIONS') {
    res.writeHead(204, CORS_HEADERS);
    res.end();
    return;
  }

  try {
    if (req.method === 'GET') {
      if (pathname === '/cubemaster' || pathname === '/cubemaster/') {
        if (existsSync(WEB_FILE)) {
          sendHtml(res, readFileSync(WEB_FILE, 'utf-8'));
        } else {
          sendHtml(res, '<h1>frontend_mock.html not found</h1>');
        }
        return;
      }

      // Serve static assets (css/, js/)
      if (pathname.startsWith('/cubemaster/css/') || pathname.startsWith('/cubemaster/js/')) {
        const relPath = pathname.slice('/cubemaster/'.length);
        serveStatic(res, relPath);
        return;
      }

      if (pathname === '/cubemaster/api/health') {
        sendJson(res, 200, { ok: true, api: BASE_API });
        return;
      }

      if (pathname === '/cubemaster/api/check-token' || pathname === '/cubemaster/api/loads') {
        const token = pickToken();
        const limit = parseInt(query.get('limit') || '1', 10);
        const { status, body } = await forward('GET', '/Loads', token, { params: { limit } });
        sendJson(res, status, body);
        return;
      }

      if (pathname.startsWith('/cubemaster/api/loads/')) {
        const token = pickToken();
        const loadId = pathname.slice('/cubemaster/api/loads/'.length);
        const { status, body } = await forward('GET', `/Loads/${loadId}`, token);
        sendJson(res, status, body);
        return;
      }

      sendJson(res, 404, { error: 'Not Found' });
      return;
    }

    if (req.method === 'POST') {
      let payload;
      try {
        payload = await readBody(req);
      } catch {
        sendJson(res, 400, { error: 'Invalid JSON body' });
        return;
      }

      if (pathname === '/cubemaster/api/loads') {
        const token = pickToken();
        const upstreamParams = {
          UOM: 'UnitMetric',
          placementsCreated: 'true',
        };
        if (query.get('graphicsCreated') === 'true') {
          upstreamParams.graphicsCreated     = 'true';
          upstreamParams.graphicsImageWidth  = query.get('graphicsImageWidth')  || '800';
          upstreamParams.graphicsImageDepth  = query.get('graphicsImageDepth')  || '600';
        }
        if (query.get('spacesCreated') === 'true') {
          upstreamParams.spacesCreated = 'true';
        }
        const { status, body } = await forward('POST', '/Loads', token, {
          params: upstreamParams,
          payload
        });
        try {
          await logLoadByFlow(query, payload, body, status);
        } catch (dbErr) {
          console.error('[DB LOG ERROR]', dbErr.message);
        }
        sendJson(res, status, body);
        return;
      }

      sendJson(res, 404, { error: 'Not Found' });
      return;
    }

    sendJson(res, 405, { error: 'Method Not Allowed' });
  } catch (err) {
    sendJson(res, 502, { error: err.message });
  }
});

server.listen(PORT, HOST, () => {
  console.log(`Local proxy running at http://${HOST}:${PORT}/cubemaster`);
  console.log('Endpoints: /cubemaster/api/health, /cubemaster/api/check-token, /cubemaster/api/loads, /cubemaster/api/loads/{id}');
});

dbPool.getConnection()
  .then((conn) => {
    conn.release();
    console.log(`[DB] Connected to MySQL ${MYSQL_HOST}:${MYSQL_PORT}/${MYSQL_DATABASE}`);
  })
  .catch((err) => {
    console.error('[DB] Connection failed:', err.message);
  });
